TensorExpand/Object detection/Mask RCNN/Mask_RCNN Exercise/详解config类.md
基础配置类。 对于自定义配置，请创建一个从这个继承的子类并重写属性那需要改变。


----------


```python
# 命名配置。 例如，'COCO'，'Experiment 3'，...等。如果你的代码需要根据哪一个做不同的事情，那么这很有用实验正在运行。

NAME = None # 子类中覆盖
# 如 NAME='shapes'
```

----------
```python
# 输入图像resing
# 图像被调整大小，使得最小的边> = IMAGE_MIN_DIM和
# 最长边是<= IMAGE_MAX_DIM。 如果两种情况都不能
# 一起满足IMAGE_MAX_DIM被执行。
IMAGE_MIN_DIM = 800
IMAGE_MAX_DIM = 1024

# 例如
IMAGE_MIN_DIM =128
IMAGE_MAX_DIM =128
```

----------

```python
# 使用的GPU数量。 对于CPU训练，请使用1
GPU_COUNT = 1
```


----------

```python
# 在每个GPU上训练的图像数量。 通常可以使用12GB的GPU处理1024x1024画面的2张图片。
# 根据您的GPU内存和图像大小进行调整。 使用最高您的GPU可以处理的编号，以获得最佳性能。
IMAGES_PER_GPU = 2

# 例如 4G GPU 128x128
# 4=128*128*n;12=1024*1024*2 ==>n=1024*1024*2*4//(128*128*12)=42
IMAGES_PER_GPU= int(1024*1024*4//(IMAGE_MAX_DIM*IMAGE_MAX_DIM*12))+1
```


----------

```python
# 每个epoch的训练步数这不需要与训练集的大小相匹配。Tensorboard更新保存在每个epoch的末尾，因此将其设置的数字越小意味着获得更频繁的TensorBoard更新。验证统计数据也会在每个epoch末期和他们计算可能需要一段时间，所以不要设置太小以免花费验证统计信息很多时间。
STEPS_PER_EPOCH = 1000

# num_images 图像数
batch_size=GPU_COUNT *IMAGES_PER_GPU
STEPS_PER_EPOCH =int(num_images/batch_size*(3/4))
```


----------

```python
# 在每个训练epoch结束时运行的验证步骤数量。更大的数字可以提高验证统计的准确性，但速度会变慢降低训练。
VALIDATION_STEPS = 50

# 例如
VALIDATION_STEPS=STEPS_PER_EPOCH//(1000//50)
```


----------

```python
# FPN金字塔每层的步幅。 这些值是基于Resnet101骨干网的。
BACKBONE_STRIDES = [4,8,16,32,64] # 与使用的神经网络相关，如果使用Resnet101则不变
```


----------

```python
# 分类类别的数量（包括背景）
NUM_CLASSES = 1 # 子类中覆盖 

# 如：
NUM_CLASSES = 1+3 # 必须包含一个背景（背景作为一个类别）
```


----------

```python
# 方形锚点的长度（以像素为单位）
RPN_ANCHOR_SCALES =（32,64,128,256,512） # 以输入图像1204x1024设置的，如果输入图像大小发生变化必须做相应的调整

# 例如
scale=1024//IMAGE_MAX_DIM
RPN_ANCHOR_SCALES = (32//scale, 64//scale, 128//scale, 256//scale, 512//scale)  # anchor side in pixels

```

----------

```python
# 每个cell处锚点的比率（宽度/高度）
# 值1表示方形锚点，0.5表示宽锚点
RPN_ANCHOR_RATIOS = [0.5,1,2] # 不变
```


----------

```python
# 锚定步伐
# 如果为1，则为骨干特征映射中的每个单元创建锚点。
# 如果是2，则为每个其他单元格创建锚点，依此类推。
RPN_ANCHOR_STRIDE = 1 # 不变

# 或
RPN_ANCHOR_STRIDE = 2
```


----------

```python
# 过滤RPN提议的非最大抑制阈值。
# 你可以在训练过程中减少这些以产生更多的提议。
RPN_NMS_THRESHOLD = 0.7  # 0.6

# 或
RPN_NMS_THRESHOLD = 0.8
```


----------

```python
# 每张图片用于RPN训练的锚点数量
RPN_TRAIN_ANCHORS_PER_IMAGE = 256

# 例如
scale=1024//IMAGE_MAX_DIM
RPN_TRAIN_ANCHORS_PER_IMAGE =256//scale

```


----------

```python
# 投资回报率在非最大压缩后保持（训练和推理）
POST_NMS_ROIS_TRAINING = 2000 # 不变
POST_NMS_ROIS_INFERENCE = 1000 # 不变
```


----------

```python
# 如果启用，则将实例掩码调整为更小的大小以减小内存负载。 建议使用高分辨率图像时使用。
USE_MINI_MASK = True
MINI_MASK_SHAPE =（56,56）＃（高度，宽度）的迷你mask

# 如
scale=1024//IMAGE_MAX_DIM
MINI_MASK_SHAPE =(56//scale,56//scale)
```


----------


----------

```python
# 如果为True，则用零填充图像，使其成为（max_dim by max_dim）
IMAGE_PADDING = True # 当前，False选项不受支持  不变
```


----------

```python
# 图像均值（RGB）
MEAN_PIXEL = np.array（[123.7，116.8，103.9]）# 不变
# 所以输入的图像需保持像素值0~255，不需事先转成0.~1.,因为会自动做归一化处理
```


----------

```python
# 每个图像的分辨率/mask头的ROI数量
# mask RCNN文件使用512，但通常RPN不会生成
# 有足够的正建议来填补这一点并保持积极的位置：否定的
# 比例为1：3。 您可以通过调整来增加投标的数量RPN网管门限。
TRAIN_ROIS_PER_IMAGE = 200

# 例如
scale=1024//IMAGE_MIN_DIM
TRAIN_ROIS_PER_IMAGE = 200//scale
```


----------

```
# 用于训练分级器/mask头的正ROI百分比
ROI_POSITIVE_RATIO = 0.33 # 不变
```


----------

```
# Pooled ROIs
POOL_SIZE = 7
MASK_POOL_SIZE = 14
MASK_SHAPE = [28,28] # 不变
```


----------

```python
# 在一个图像中使用的地面实况实例的最大数量
MAX_GT_INSTANCES = 100 # 不变
```


----------

```python
# RPN和最终检测的边界框细化标准差。
RPN_BBOX_STD_DEV = np.array（[0.1,0.1,0.2,0.2]） # 不变
BBOX_STD_DEV = np.array（[0.1,0.1,0.2,0.2]） # 不变
```


----------

```python
# 最大检测次数
DETECTION_MAX_INSTANCES = 100 # 不变

# 例如
scale=1024//IMAGE_MIN_DIM
DETECTION_MAX_INSTANCES =100*scale*2//3
```


----------

```python
# 接受检测到的实例的最小概率值低于此阈值的ROI将被跳过
DETECTION_MIN_CONFIDENCE = 0.7

# 例如
DETECTION_MIN_CONFIDENCE = 0.8
```


----------

```python
# 检测的非最大抑制阈值
DETECTION_NMS_THRESHOLD = 0.3 # 不变
```


----------

```python
# 学习效率和动力
# Mask RCNN文件使用lr = 0.02，但在TensorFlow上引起权重爆炸。 可能由于优化器的差异执行。
LEARNING_RATE = 0.001  # 不变
LEARNING_MOMENTUM = 0.9 # 不变
```


----------

```python
# 重量衰减正则化
WEIGHT_DECAY = 0.0001 # 不变
```


----------

```python
# 使用RPN投资回报率或外部生成的投资回报率进行训练
# 在大多数情况下保持真实。 如果您想训练，请设置为False
# 由代码生成的ROI分支，而不是来自ROI的ROIRPN。 例如，调试分类器头，而不必训练RPN。
USE_RPN_ROIS = True # 不变
```


----------




TensorExpand/Object detection/Mask RCNN/matterport-Mask_RCNN/config.py
"""
Mask R-CNN
Base Configurations class.
Copyright (c) 2017 Matterport, Inc.
Licensed under the MIT License (see LICENSE for details)
Written by Waleed Abdulla
"""

import math
import numpy as np


# Base Configuration Class
# Don't use this class directly. Instead, sub-class it and override
# the configurations you need to change.

class Config(object):
    """Base configuration class. For custom configurations, create a
    sub-class that inherits from this one and override properties
    that need to be changed.
    """
    # Name the configurations. For example, 'COCO', 'Experiment 3', ...etc.
    # Useful if your code needs to do things differently depending on which
    # experiment is running.
    NAME = None  # Override in sub-classes

    # NUMBER OF GPUs to use. For CPU training, use 1
    GPU_COUNT = 1

    # Number of images to train with on each GPU. A 12GB GPU can typically
    # handle 2 images of 1024x1024px.
    # Adjust based on your GPU memory and image sizes. Use the highest
    # number that your GPU can handle for best performance.
    IMAGES_PER_GPU = 2

    # Number of training steps per epoch
    # This doesn't need to match the size of the training set. Tensorboard
    # updates are saved at the end of each epoch, so setting this to a
    # smaller number means getting more frequent TensorBoard updates.
    # Validation stats are also calculated at each epoch end and they
    # might take a while, so don't set this too small to avoid spending
    # a lot of time on validation stats.
    STEPS_PER_EPOCH = 1000

    # Number of validation steps to run at the end of every training epoch.
    # A bigger number improves accuracy of validation stats, but slows
    # down the training.
    VALIDATION_STEPS = 50

    # The strides of each layer of the FPN Pyramid. These values
    # are based on a Resnet101 backbone.
    BACKBONE_STRIDES = [4, 8, 16, 32, 64]

    # Number of classification classes (including background)
    NUM_CLASSES = 1  # Override in sub-classes

    # Length of square anchor side in pixels
    RPN_ANCHOR_SCALES = (32, 64, 128, 256, 512)

    # Ratios of anchors at each cell (width/height)
    # A value of 1 represents a square anchor, and 0.5 is a wide anchor
    RPN_ANCHOR_RATIOS = [0.5, 1, 2]

    # Anchor stride
    # If 1 then anchors are created for each cell in the backbone feature map.
    # If 2, then anchors are created for every other cell, and so on.
    RPN_ANCHOR_STRIDE = 1

    # Non-max suppression threshold to filter RPN proposals.
    # You can reduce this during training to generate more propsals.
    RPN_NMS_THRESHOLD = 0.7

    # How many anchors per image to use for RPN training
    RPN_TRAIN_ANCHORS_PER_IMAGE = 256

    # ROIs kept after non-maximum supression (training and inference)
    POST_NMS_ROIS_TRAINING = 2000
    POST_NMS_ROIS_INFERENCE = 1000

    # If enabled, resizes instance masks to a smaller size to reduce
    # memory load. Recommended when using high-resolution images.
    USE_MINI_MASK = True
    MINI_MASK_SHAPE = (56, 56)  # (height, width) of the mini-mask

    # Input image resing
    # Images are resized such that the smallest side is >= IMAGE_MIN_DIM and
    # the longest side is <= IMAGE_MAX_DIM. In case both conditions can't
    # be satisfied together the IMAGE_MAX_DIM is enforced.
    IMAGE_MIN_DIM = 800
    IMAGE_MAX_DIM = 1024
    # If True, pad images with zeros such that they're (max_dim by max_dim)
    IMAGE_PADDING = True  # currently, the False option is not supported

    # Image mean (RGB)
    MEAN_PIXEL = np.array([123.7, 116.8, 103.9])

    # Number of ROIs per image to feed to classifier/mask heads
    # The Mask RCNN paper uses 512 but often the RPN doesn't generate
    # enough positive proposals to fill this and keep a positive:negative
    # ratio of 1:3. You can increase the number of proposals by adjusting
    # the RPN NMS threshold.
    TRAIN_ROIS_PER_IMAGE = 200

    # Percent of positive ROIs used to train classifier/mask heads
    ROI_POSITIVE_RATIO = 0.33

    # Pooled ROIs
    POOL_SIZE = 7
    MASK_POOL_SIZE = 14
    MASK_SHAPE = [28, 28]

    # Maximum number of ground truth instances to use in one image
    MAX_GT_INSTANCES = 100

    # Bounding box refinement standard deviation for RPN and final detections.
    RPN_BBOX_STD_DEV = np.array([0.1, 0.1, 0.2, 0.2])
    BBOX_STD_DEV = np.array([0.1, 0.1, 0.2, 0.2])

    # Max number of final detections
    DETECTION_MAX_INSTANCES = 100

    # Minimum probability value to accept a detected instance
    # ROIs below this threshold are skipped
    DETECTION_MIN_CONFIDENCE = 0.7

    # Non-maximum suppression threshold for detection
    DETECTION_NMS_THRESHOLD = 0.3

    # Learning rate and momentum
    # The Mask RCNN paper uses lr=0.02, but on TensorFlow it causes
    # weights to explode. Likely due to differences in optimzer
    # implementation.
    LEARNING_RATE = 0.001
    LEARNING_MOMENTUM = 0.9

    # Weight decay regularization
    WEIGHT_DECAY = 0.0001

    # Use RPN ROIs or externally generated ROIs for training
    # Keep this True for most situations. Set to False if you want to train
    # the head branches on ROI generated by code rather than the ROIs from
    # the RPN. For example, to debug the classifier head without having to
    # train the RPN.
    USE_RPN_ROIS = True

    def __init__(self):
        """Set values of computed attributes."""
        # Effective batch size
        self.BATCH_SIZE = self.IMAGES_PER_GPU * self.GPU_COUNT

        # Input image size
        self.IMAGE_SHAPE = np.array(
            [self.IMAGE_MAX_DIM, self.IMAGE_MAX_DIM, 3])

        # Compute backbone size from input image size
        self.BACKBONE_SHAPES = np.array(
            [[int(math.ceil(self.IMAGE_SHAPE[0] / stride)),
              int(math.ceil(self.IMAGE_SHAPE[1] / stride))]
             for stride in self.BACKBONE_STRIDES])

    def display(self):
        """Display Configuration values."""
        print("\nConfigurations:")
        for a in dir(self):
            if not a.startswith("__") and not callable(getattr(self, a)):
                print("{:30} {}".format(a, getattr(self, a)))
        print("\n")
